scrapy_poet/registry.py

```diff
@@ -0,0 +1,149 @@
+import logging
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Any, Callable, Dict, Iterable, Mapping, Optional, Type
+from warnings import warn
+
+from scrapy import Request
+from scrapy.crawler import Crawler
+from url_matcher import URLMatcher
+from web_poet import ItemPage, RulesRegistry
+from web_poet.rules import ApplyRule
+
+from scrapy_poet.utils._deprecation import _create_deprecated_class
+
+logger = logging.getLogger(__name__)
+
+
+class OverridesRegistryBase(ABC):
+    @abstractmethod
+    def overrides_for(self, request: Request) -> Mapping[Callable, Callable]:
+        """
+        Return a ``Mapping`` (e.g. a ``dict``) with type translation rules.
+        The key is the source type that is wanted to be replaced by
+        value, which is also a type.
+        """
+        pass
+
+
+class OverridesAndItemRegistry(OverridesRegistryBase, RulesRegistry):
+    """This registry implements the functionalities for returning the override
+    for a given request as well as returning the page object capable of returning
+    an item class for a given URL.
+
+    Read the :ref:`rules-from-web-poet` tutorial for more information.
+
+    It reads the rules from the ``SCRAPY_POET_RULES`` setting which is a list of
+    :py:class:`web_poet.rules.ApplyRule` instances:
+
+    .. code-block:: python
+
+        from web_poet import default_registry
+
+        SCRAPY_POET_RULES = default_registry.get_rules()
+
+    .. _web-poet: https://web-poet.readthedocs.io
+
+    Calling ``default_registry.get_rules()`` finds all the rules annotated using
+    web-poet_'s :py:func:`web_poet.handle_urls` as a decorator that were registered
+    into ``web_poet.default_registry`` (an instance of
+    :py:class:`web_poet.rules.RulesRegistry`).
+
+    .. warning::
+
+        The :py:func:`web_poet.rules.consume_modules` must be called first
+        if you need to properly import rules from other packages. For example:
+
+        .. code-block:: python
+
+            from web_poet import default_registry, consume_modules
+
+            consume_modules("external_package_A.po", "another_ext_package.lib")
+            SCRAPY_POET_RULES = default_registry.get_rules()
+
+    """
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler) -> Crawler:
+        if "SCRAPY_POET_OVERRIDES" in crawler.settings:
+            msg = (
+                "The SCRAPY_POET_OVERRIDES setting is deprecated. "
+                "Use SCRAPY_POET_RULES instead."
+            )
+            warn(msg, DeprecationWarning, stacklevel=2)
+            rules = crawler.settings.getlist("SCRAPY_POET_OVERRIDES", [])
+        else:
+            rules = crawler.settings.getlist("SCRAPY_POET_RULES", [])
```

Comment:
I wonder, for these deprecated settings, if we should consider a scenario where both settings are declared. Currently, the properly-named setting is ignored. Maybe the old-named setting should be ignored instead, in that scenario.

Reply:
Good call! Updated to prioritize the new setting first in https://github.com/scrapinghub/scrapy-poet/pull/103/commits/41412398440ed3a517bb3040483390c16cdbd479